pkg/server/server.go

```diff
@@ -855,6 +855,24 @@ func (s *Server) ShowTxnList() []*txninfo.TxnInfo {
 	return rs
 }
 
+// UpdateProcessCPUTime updates specific process's tidb CPU time when the process is still running
+// It implements ProcessCPUTimeUpdater interface
+func (s *Server) UpdateProcessCPUTime(connID uint64, sqlID uint64, cpuTime time.Duration) {
+	s.rwlock.RLock()
+	conn, ok := s.clients[connID]
+	s.rwlock.RUnlock()
+	if !ok || conn.ctx.GetSessionVars().SQLID.Load() != sqlID {
+		return
+	}
+	vars := conn.ctx.GetSessionVars()
+	if vars != nil {
+		sc := vars.StmtCtx
```

Comment:
Data Race？https://github.com/pingcap/tidb/blob/master/pkg/executor/executor.go#L2096

Reply:
Nice, I'll check it.

Reply:
It seems hard to add protection for accessing statement context. Besides, using statement context may fail to track some tidb cpu time, like parsing time, thus move cpu usage to SessionVariables.